src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -15,6 +15,8 @@ namespace System.Text.Json.Serialization.Metadata
     [DebuggerDisplay("ConverterStrategy.{ConverterStrategy}, {Type.Name}")]
     public partial class JsonTypeInfo
     {
+        internal const string JsonObjectTypeName = "System.Text.Json.Node.JsonObject";
```

Comment:
Should we add tests to guard against custom user types with the same full name?

Reply:
Do we do this elsewhere to your knowledge? Note I added an assembly check to prevent this.

Reply:
We wanted to add such tests for immutable collection (de)serialization since it is based on strings and reflection, but ran into issues because we couldn't force custom types to have the right assembly name (since the test assembly is `System.Text.Json.Tests`). Okay to ignore since the intent is almost certainly malicious if the type has the same assembly name.